File: src/test/resources/org/bau/benchmarks/go/linkedList.go
// https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/blob/master/2026/02/13/stress_gc.go

package main

import (
    "fmt"
    "time"
    "runtime"
)

type Node struct {
    value int
    next  *Node
}

func createLinkedList(num int) *Node {
    head := &Node{value: 0}
    current := head
    for i := 1; i < num; i++ {
        current.next = &Node{value: i}
        current = current.next
    }
    return head
}

func main() {
    runtime.GOMAXPROCS(1)
    fmt.Println("Starting GC stress test...")
    maxdiff := time.Duration(0)

    objects1 := createLinkedList(50_000_000)
    start := time.Now()
    dummy := 0

    // Create a lot of cyclic garbage
    for batch := 0; batch < 1_000_000; batch++ {
        batchStart := time.Now()
        // compute the delay between batchStart and start
        diff := batchStart.Sub(start)
        if diff > maxdiff {
            maxdiff = diff
        }
        start = batchStart
        objects := createLinkedList(1000)
        dummy += objects.value
    }
    fmt.Printf("Max delay between batch start and overall start: %v\n", maxdiff)
    fmt.Printf("GC stress test completed. Value: %d %d\n", objects1.value, dummy)
}

File: src/test/resources/org/bau/benchmarks/python/linkedList.py
# https://lemire.me/blog/2026/02/15/how-bad-can-python-stop-the-world-pauses-get/

import time

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    def add_next(self, node):
        self.next = node

def create_linked_list(limit):
    """ create a linked list of length 'limit' """
    head = Node(0)
    current = head
    for i in range(1, limit):
        new_node = Node(i)
        current.add_next(new_node)
        current = new_node
    return head

def main():
    print("Starting loading linked list...")
    x = create_linked_list(50_000_000)
    print("Initial batch completed, starting stress test...")
    start = time.time()
    worse_case = 0.0
    for i in range(1_000_000):
        batch_start = time.time()
        if batch_start - start > worse_case:
            worse_case = batch_start - start
        start = batch_start
        create_linked_list(1000)
    print(f"Max delay between batch start and overall start: {worse_case * 1000:.4f} milliseconds", flush=True)

main()